### PR TITLE
shell args

### PR DESCRIPTION
**Pull request by p0lyn0mial**
This PR provides convenient functions for dealing with command-line arguments.

### CODE REVIEW

**Comment by p0lyn0mial** [comment]
/hold

need unit test for `ToShellEscapedArguments`

File: pkg/operator/v1helpers/args.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+// shellEscapePattern determines if a string should be enclosed in single quotes
+// so that it can safely be passed to shell command line.
+var shellEscapePattern *regexp.Regexp
+
+func init() {
+	// some characters have special meaning to the shell and may need to be escaped
+	// the following regexp defines safe characters that don't require escaping
```
Comment (sttts) [inline pkg/operator/v1helpers/args.go:18]: where is this from?

File: pkg/operator/v1helpers/args.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+func ToShellEscapedArguments(unstructuredArgs map[string]interface{}) (map[string][]string, error) {
+	return toStructuredArguments(unstructuredArgs, true)
+}
+
+// ToStructuredArguments process the unstructured arguments usually retrieved from an operator's configuration file under a specific key.
+// There are only two supported/valid types for arguments, that is []sting or string.
+// Passing a different type yeld an error.
+//
+// Use ToFlagSlice function to get a slice of string flags.
+func ToStructuredArguments(unstructuredArgs map[string]interface{}) (map[string][]string, error) {
```
Comment (sttts) [inline pkg/operator/v1helpers/args.go:35]: `FlagsFromUnstructured` would say more, no?

File: pkg/operator/v1helpers/args.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+				return nil, fmt.Errorf("unable to process an argument, incorrect value %v under %v key, expected []string or string", argRawValue, argName)
+			}
+			argsSlice = append(argsSlice, str)
+		}
+
+		escapedArgsSlice := make([]string, len(argsSlice))
+		for index, str := range argsSlice {
+			arg := str
+			if escape {
+				arg = maybeQuote(str)
```
Comment (sttts) [inline pkg/operator/v1helpers/args.go:93]: escaping here is at the wrong spot, isn't it? You should have a `ToShellFlagString` func that does the escaping.

Comment (sttts) [reply pkg/operator/v1helpers/args.go:93]: ... and that only returns one string.

Comment (p0lyn0mial) [reply pkg/operator/v1helpers/args.go:93]: good catch, thx!

File: pkg/operator/v1helpers/args.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	// some characters have special meaning to the shell and may need to be escaped
+	// the following regexp defines safe characters that don't require escaping
+	shellEscapePattern = regexp.MustCompile(`[^\w@%+=:,./-]`)
+}
+
+// FlagsFromUnstructuredEscaped process the given arguments and determines if they should be escaped so that they can be safely passed as shell command line arguments.
+// unstructuredArgs holds unprocessed arguments usually retrieved from an operator's configuration file under a specific key.
+//
+// Use ToFlagSlice function to get a slice of string flags.
+func FlagsFromUnstructuredEscaped(unstructuredArgs map[string]interface{}) (map[string][]string, error) {
```
Comment (sttts) [inline pkg/operator/v1helpers/args.go:26]: I still think this is the wrong spot for escaping. Escape when turning a `[]string` into *one* string.

File: pkg/operator/v1helpers/args.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+// shellEscapePattern determines if a string should be enclosed in single quotes
+// so that it can safely be passed to shell command line.
+var shellEscapePattern *regexp.Regexp
+
+func init() {
+	// some characters have special meaning to the shell and may need to be escaped
+	// the following regexp defines safe characters that don't require escaping
+	shellEscapePattern = regexp.MustCompile(`[^\w@%+=:,./-]`)
```
Comment (mfojtik) [inline pkg/operator/v1helpers/args.go:19]: this looks strange, why is this in init and not in var declaration?

File: pkg/operator/v1helpers/args.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+func FlagsFromUnstructured(unstructuredArgs map[string]interface{}) (map[string][]string, error) {
+	return flagsFromUnstructured(unstructuredArgs, false)
+}
+
+// ToFlagSlice transforms the provided arguments to a slice of string flags.
+// A flag name is taken directly from the key and the value is simply attached.
+// A flag is repeated iff it has more than one value.
+func ToFlagSlice(args map[string][]string) []string {
+	var keys []string
+	for key := range args {
```
Comment (mfojtik) [inline pkg/operator/v1helpers/args.go:44]: won't `keys := append([]string{}, args...)` do the same?

File: pkg/operator/v1helpers/args.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+		keys = append(keys, key)
+	}
+	sort.Strings(keys)
+
+	var flags []string
+	for _, key := range keys {
+		for _, token := range args[key] {
+			flags = append(flags, fmt.Sprintf("--%s=%v", key, token))
```
Comment (mfojtik) [inline pkg/operator/v1helpers/args.go:52]: s/%v/%s/ ?

File: pkg/operator/v1helpers/args.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+		return "''"
+	}
+	if shellEscapePattern.MatchString(s) {
+		return "'" + strings.Replace(s, "'", "'\"'\"'", -1) + "'"
+	}
+
+	return s
+}
+
+func flagsFromUnstructured(unstructuredArgs map[string]interface{}, escape bool) (map[string][]string, error) {
```
Comment (mfojtik) [inline pkg/operator/v1helpers/args.go:73]: nit: this is long and complicated enough to deserve one-line comment :-)

File: pkg/operator/v1helpers/args.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	ret := map[string][]string{}
+	for argName, argRawValue := range unstructuredArgs {
+		var argsSlice []string
```
Comment (mfojtik) [inline pkg/operator/v1helpers/args.go:76]: nit: group this?

File: pkg/operator/v1helpers/args.go (hunk 6 above)
Comment (p0lyn0mial) [reply pkg/operator/v1helpers/args.go:44]: `args` is a map, we need to iterate to get the keys.

**Pull request merged**